### ISSUE TITLE
[LAMP] <write permission errors on the /opt/bitnami/apache/htdocs folder>

### ISSUE THREAD

**Issue by linuxguist** [bitnami/vms#604]
### Describe your suggestion as much as you can

Dear Bitnami Support,

First of all, thank you very much for your contribution of Stacks and Virtual Machines to the World !
It is very helpful indeed.

Recently, I downloaded the latest LAMP VM for a new PHP project of mine and discovered ( through lots of Google Researches ) that the write permission errors on the filesystem, can be easily resolved, as follows: ( This might probably apply to your other VMs as well. Offcourse, take whatever inputs you find it applicable from this )

--------------------------------------------------------

To prevent write permission errors on the /opt/bitnami/apache/htdocs or the /var/www/html folder, the background service user of php/apache service (either www-data or daemon user/group) must be assigned certain privileges on these folders, as below:

1. Change the primary group of our frontend user first and also keep its current group, as secondary:

sudo usermod -g daemon bitnami

sudo usermod -a -G bitnami bitnami

sudo reboot

id -a

2. Change directory and set sticky bit for the group and also assign group and other permissions:

cd /opt/bitnami/apache/htdocs  # or cd /var/www/html folder

sudo chmod g+s .

sudo setfacl -d -m g::rwx .
sudo setfacl -d -m o::rx .

sudo getfacl .

sudo setfacl -R -m g::rwx .

That's all. Now our web applications must be able to comfortably write to the document root folder.

As a bonus, if you set the error_log variable to some file name in php.ini, then this will also get written in our web application folder and will serve our needs for debugging / troubleshooting purposes.

Many web hosting providers set the value for this variable in php.ini file.

--------------------------------------------------------

Hope this also improves the RATINGS given by your users for all these VMs.

Thanks and Best Regards,

Nathan S.R.

**Comment by jotamartos** [comment]
Hi @linuxguist,

Thank you for your kind words and for the information you posted. 

I'd like to tell you a different way to solve that problem. Once you copy your project's files to the /opt/bitnami/apache/htdocs folder, please run these commands to allow Apache (daemon) read and write there.

```
sudo chown -R bitnami:daemon /opt/bitnami/apache/htdocs
sudo chmod -R g+w /opt/bitnami/apache/htdocs
```

That's all you need. Have a good week!

**Comment by linuxguist** [comment]
Hi @jotamartos,

Thank you for the quick reply.

Actually, this issue is different. Each time we transfer new files/folders through the FileZilla program, as the Bitnami user, the commands that you said, need to be repeated, by even less knowledgable users.

The solution that I proposed is a one time fix, forever in all the VMs that you host.

A humble request to you, is that please execute these commands at your end, on a test machine, and see the difference it makes when transferring new files/folders. No more permission problems, to anyone, anymore !!!!

Also, make a mock syntax error in any php file and see if it generates an error_log in that same folder, before and after doing the changes that I suggested. Without seeing their own error log, people complain about bitnami VM not working, is a poor excuse indeed !

Regards,

Nathan S.R.

**Comment by jotamartos** [comment]
I see your point and it's completely valid. However, we prefer to keep the permissions configuration as it's now. Users will need to configure the permissions of the files they copy (please note that there may be some files that do not need write permissions due to security reasons). If they are struggling with permissions issues, they can follow this post to configure the permissions your way and test the application.

**Comment by mimoudix** [comment]
@jotamartos  Im using SamKirkland/web-deploy@v1 using rsync and it can't sync files unless i give htdocs the following permissions :

```
sudo chown -R bitnami:daemon /opt/bitnami/apache/htdocs
sudo chmod -R g+w /opt/bitnami/apache/htdocs
```

Is this concidered a bad/insecure practice ?

im using a cron since I only need permissions on logs folder and Uploads. The rest os not important for me

```
# Every 15 mins give apache permission to App/logs and content/uploads
*/15 * * * * sudo chown -R bitnami:daemon /opt/bitnami/apache/htdocs/App/logs && sudo chmod -R g+w /opt/bitnami/apache/htdocs/App/logs && sudo chown -R bitnami:daemon /opt/bitnami/apache/htdocs/content/uploads && sudo chmod -R g+w /opt/bitnami/apache/htdocs/content/uploads
```

**Comment by jotamartos** [comment]
>Is this concidered a bad/insecure practice ?

No, that's correct. If the application needs write permissions in the folder, you need to provide it with the necessary permissions to work. 

>im using a cron since I only need permissions on logs folder and Uploads. The rest os not important for me

That's perfect! You should just give write permissions to the folders that require it. However, you shouldn't need to execute this every 15 mins, the permissions shouldn't change once you run this once.
